### PR TITLE
Add schema support

### PR DESCRIPTION
**Pull request by Yomanz**
Closes https://github.com/jasync-sql/jasync-sql/issues/258

I normally do Scala, not Kotlin so I apologise for any mistakes.

### CODE REVIEW

File: db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt [hunk 1]
```diff
@@ -53,7 +53,8 @@ data class Configuration @JvmOverloads constructor(
     val applicationName: String? = null,
     val interceptors: List<Supplier<QueryInterceptor>> = emptyList(),
     val eventLoopGroup: EventLoopGroup = NettyUtils.DefaultEventLoopGroup,
-    val executionContext: Executor = ExecutorServiceUtils.CommonPool
+    val executionContext: Executor = ExecutorServiceUtils.CommonPool,
+    val currentSchema: String? = null,
```
Comment (oshai) [inline db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt:57]: maybe just call the parameter "schema"?

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/util/URLParserSpec.kt [hunk 2]
```diff
@@ -86,6 +86,19 @@ class URLParserSpec {
         assertThat(configuration.ssl.mode).isEqualTo(SSLConfiguration.Mode.VerifyFull)
     }
 
+    @Test
+    fun `postgresql URLParser should  create a connection , schema defined`() {
```
Comment (oshai) [inline postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/util/URLParserSpec.kt:90]: can you please also add a test that uses the schema (select on a table without qualifying the table by schema)?

**Review by oshai — COMMENTED** [review]
Thanks for the PR!

File: db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt (hunk 1 above)
Comment (Yomanz) [reply db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt:57]: It's `currentSchema` on the URL so I think it makes sense to keep it as that.

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/util/URLParserSpec.kt (hunk 2 above)
Comment (Yomanz) [reply postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/util/URLParserSpec.kt:90]: I'll have a go later today.

**Comment by Yomanz** [comment]
> > Sorry for the lack of response, I'm not really sure what you want me to do for the test.
> 
> I think a test that is doing a query and use the search path - means add a schema to search path and then query that schema without it's qualified name.

But how would I be able to see whether it's doing it on the right schema anyway?

**Comment by oshai** [comment]
> > > Sorry for the lack of response, I'm not really sure what you want me to do for the test.
> > 
> > 
> > I think a test that is doing a query and use the search path - means add a schema to search path and then query that schema without it's qualified name.
> 
> But how would I be able to see whether it's doing it on the right schema anyway?

If you have 2 schemas (default + on the search path) and your query the one on the search path, isn't that proves it's on the search path?

**Comment by oshai** [comment]
> > > > > Sorry for the lack of response, I'm not really sure what you want me to do for the test.
> > > > 
> > > > 
> > > > I think a test that is doing a query and use the search path - means add a schema to search path and then query that schema without it's qualified name.
> > > 
> > > 
> > > But how would I be able to see whether it's doing it on the right schema anyway?
> > 
> > 
> > If you have 2 schemas (default + on the search path) and your query the one on the search path, isn't that proves it's on the search path?
> 
> True, have you got any existing tests I can partially copy?

Take a look at PreparedStatementSpec. In mysql there is also QuerySpec, so I guess you can create a QuerySpec class.

**Comment by Yomanz** [comment]
Any update?

**Comment by oshai** [comment]
hi,
I am trying to find a contributor to write some tests, but it might take some time.
Is there any urgency from your side?

**Comment by Yomanz** [comment]
Not necessarily urgent but I'm currently using a hacked together version of quill and the old SQL library (that this replaced) so I can't update quill easily etc.

**Comment by Yomanz** [comment]
> hi, I am trying to find a contributor to write some tests, but it might take some time. Is there any urgency from your side?

Any luck, It's becoming a bit more urgent now.

**Comment by Yomanz** [comment]
@oshai Any update on this?

**Comment by oshai** [comment]
> @oshai Any update on this?

sorry, nothing new at the moment.

**Comment by Yomanz** [comment]
@oshai I've attempted to add the test, are you able to check it and let me know if it's ok.

**Comment by oshai** [comment]
I will take a look.

**Comment by oshai** [comment]
yes, looks ok. is it possible to add a test for table not on the search path that is not found? throws an error?

**Comment by Yomanz** [comment]
> yes, looks ok. is it possible to add a test for table not on the search path that is not found? throws an error?

Sorry for the delay on this, can you let me know if that looks right.
I wasn't sure what the exception would actually be so if that's wrong could you point me to the right one?

**Comment by Yomanz** [comment]
@oshai Hi, any updates?

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/QuerySpec.kt [hunk 3]
```diff
@@ -27,4 +29,21 @@ class QuerySpec : DatabaseTestHelper() {
...
+        withHandler(ContainerHelper.defaultConfiguration.copy(currentSchema = "testing")) { handler ->
+            executeDdl(handler, this.messagesCreate)
+            executeDdl(handler, "INSERT INTO messages (content) VALUES ('Hello')")
+        }
+
+        withHandler { handler ->
+            val rows = executePreparedStatement(handler, "SELECT id FROM messages").rows
+            assertThat(rows.length).isEqualTo(1)
+
+            verifyException(ExecutionException::class.java) {
```
Comment (Yomanz) [inline postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/QuerySpec.kt:43]: I don't think this is the right exception but I'm not sure what to do for it.

**Comment by oshai** [comment]
Sorry, I am a bit busy, but will try to review it soon.

**Comment by Yomanz** [comment]
> Sorry, I am a bit busy, but will try to review it soon.

That'd be great, thanks!

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/QuerySpec.kt (hunk 3 above)
Comment (oshai) [reply postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/QuerySpec.kt:43]: you can do something like this to check the inner exception: verifyException(ExecutionException::class.java, IllegalStateException::class.java)

File: db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt (hunk 1 above)
Comment (oshai) [inline db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt:57]: maybe add a comment that this only supported in postgres at the moment?

**Review by oshai — COMMENTED** [review]
I can also submit it as is, let me know if you want to do more changes for this cl. sorry again for delays.

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/QuerySpec.kt (hunk 3 above)
Comment (Yomanz) [reply postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/QuerySpec.kt:43]: I'm not sure what the inner exception would be so best to just leave as is for now.

**Comment by Yomanz** [comment]
> I can also submit it as is, let me know if you want to do more changes for this cl. sorry again for delays.

That'd be great, thanks!

**Comment by oshai** [comment]
Thanks again!

**Comment by Yomanz** [comment]
> Thanks again!

Happy to help, looking forward to it being released.

**Comment by oshai** [comment]
I found out that ci was not running properly on the PRs so making some fixes.

**Comment by Yomanz** [comment]
@oshai How am I able to run tests locally, not used to the Kotlin ecosystem.

**Comment by oshai** [comment]
you can either run tests from intellij or from cli: `./gradlew :postgresql-async:test`

logs are at `postgresql-async/target/postgresql-async-tests.log`

**Comment by oshai** [comment]
created 279 to fix the tests.

**Comment by oshai** [comment]
release in 2.0.5.

**Pull request merged**